src/V3Randomize.cpp

```diff
@@ -1520,11 +1574,12 @@ class RandomizeVisitor final : public VNVisitor {
             if (memberVarp->user3()) return;  // Handled in constraints
             const AstNodeDType* const dtypep = memberVarp->dtypep()->skipRefp();
             if (VN_IS(dtypep, BasicDType) || VN_IS(dtypep, StructDType)
-                || VN_IS(dtypep, UnionDType)) {
+                || VN_IS(dtypep, UnionDType) || VN_IS(dtypep, PackArrayDType)
+                || VN_IS(dtypep, UnpackArrayDType) || VN_IS(dtypep, DynArrayDType)) {
                 AstVar* const randcVarp = newRandcVarsp(memberVarp);
                 AstVarRef* const refp = new AstVarRef{fl, classp, memberVarp, VAccess::WRITE};
                 AstNodeStmt* const stmtp = newRandStmtsp(fl, refp, randcVarp);
-                basicRandomizep->addStmtsp(stmtp);
+                basicRandomizep->addStmtsp(new AstBegin{fl, "", stmtp});
```

Comment:
Does this solve problems with randomizing multiple arrays (or deep nested arrays, arrays of queues, queues of arrays)? Perhaps unique names can be made redundant by creating per-foreach AstBegin blocks.

Reply:
> Does this solve problems with randomizing multiple arrays (or deep nested arrays, arrays of queues, queues of arrays)? Perhaps unique names can be made redundant by creating per-foreach AstBegin blocks.

I think, It solves this problem https://github.com/verilator/verilator/pull/5415#issuecomment-2334053124.

Sorry, I don't get this. Could you please explain more about "Perhaps unique names can be made redundant by creating per-foreach AstBegin blocks."?

Thank you.

Reply:
I meant that it might solve some of the issues if begin blocks are created together with foreach loops, therefore `new AstForeach{...}` would become `new AstBegin{..., new AstForeach{...}, ...}` (and then without the one big AstBegin here).

I also meant that it is also possible that unique names would not be needed then (Begin blocks would ensure separate scopes are created for the created variables).

Reply:
I had previously attempted to combine begin and foreach into a single `stmtsp `to return, but there was a type mismatch which required changing the return type of the entire `newRandStmtsp`. Additionally, combining them caused an issue where the foreach loop was automatically expanded into a while loop. This affects the randomization of all arrays except dynamic arrays, and I believe it may be related to the use of the `at write` AstCmethod for dynamic arrays.

For now, I suggest we proceed with this approach. Once I finish supporting all array types (only associative arrays and queues are left), we can revisit and discuss how to optimize this point. Would that be alright with you?

Reply:
I am also fine with the way it is currently implemented.